core/src/test/java/kafka/zk/ZkBrokerRegistrationTest.java

```diff
@@ -0,0 +1,298 @@
+ * the use case.
+ */
+public class ZkBrokerRegistrationTest {
+    private static final Logger log = LoggerFactory.getLogger(ZkBrokerRegistrationTest.class);
+
+    private KafkaConfig kafkaConfig;
+    private BrokerInfo brokerInfo;
+    private int zkPort;
+
+    private class SandboxedZookeeper extends Thread {
+        private final CountDownLatch zookeeperStopLatch;
+        private final CountDownLatch zookeeperStartLatch = new CountDownLatch(1);
+        private final ZkTestContext spec;
+        private InstrumentedRequestProcessor processor;
+
+        SandboxedZookeeper(CountDownLatch zookeeperStopLatch, ZkTestContext spec) {
+            this.zookeeperStopLatch = zookeeperStopLatch;
+            this.spec = spec;
+        }
+
+        public void run() {
+            ServerCnxnFactory cnxnFactory = null;
+
+            try {
+                Path dataDir = Files.createTempDirectory("zk");
+
+                Properties zkProperties = new Properties();
+                zkProperties.put("dataDir", dataDir.toFile().getPath());
+                zkProperties.put("clientPort", String.valueOf(zkPort));
+                zkProperties.put("serverCnxnFactory", "org.apache.zookeeper.server.NettyServerCnxnFactory");
+
+                QuorumPeerConfig config = new QuorumPeerConfig();
+                config.parseProperties(zkProperties);
+                FileTxnSnapLog txnLog = new FileTxnSnapLog(config.getDataLogDir(), config.getDataDir());
+
+                ZooKeeperServer zookeeper = new InstrumentedZooKeeperServer(
+                    null,
+                    txnLog,
+                    config.getTickTime(),
+                    config.getMinSessionTimeout(),
+                    config.getMaxSessionTimeout(),
+                    config.getClientPortListenBacklog(),
+                    null,
+                    config.getInitialConfig(),
+                    spec) {
+
+                    @Override
+                    protected void setupRequestProcessors() {
+                        processor = new InstrumentedRequestProcessor(this, spec);
+                        RequestProcessor syncProcessor = new SyncRequestProcessor(this, processor);
+                        ((SyncRequestProcessor) syncProcessor).start();
+                        firstProcessor = new PrepRequestProcessor(this, syncProcessor);
+                        ((PrepRequestProcessor) firstProcessor).start();
+                    }
+                };
+
+                cnxnFactory = ServerCnxnFactory.createFactory();
+                cnxnFactory.configure(
+                    config.getClientPortAddress(),
+                    config.getMaxClientCnxns(),
+                    config.getClientPortListenBacklog(),
+                    false);
+                cnxnFactory.startup(zookeeper);
+
+                zookeeperStartLatch.countDown();
+                zookeeperStopLatch.await();
```

Comment:
Hmm, at least in theory, the use of `Thread#stop()` is [discouraged](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/doc-files/threadPrimitiveDeprecation.html). Its Javadoc indicates:

> *Deprecated: This method is inherently unsafe. 
> [...]
> Many uses of stop should be replaced by code that simply modifies some variable to indicate that the target thread should stop running. The target thread should check this variable regularly, and return from its run method in an orderly fashion if the variable indicates that it is to stop running. If the target thread waits for long periods (on a condition variable, for example), the interrupt method should be used to interrupt the wait.*

I replaced the count-down latch with a flag-and-interrupt approach and a closeable pattern borrowed from e.g. the Kafka client.

Reply:
Indeed, I didn't mean to suggest `Thread#stop()`. 

Currently my understanding is that `SandoxedZookeeper` is a thread that when run, starts ZK (asynchronously) and then that same thread just actively waits for a signal before it shutdowns ZK.

What I'm proposing is that `SandoxedZookeeper` not be a `Thread`, and provide a `startZk()` (or just `start()` method that does what is currently being done in `Thread#run()` to bring up ZK, and also provide a `stop()` method to do what is currently being done after the signal is received. i.e. instead of propagating a signal, we could just invoke `stop()`. 

I'm hoping this way we can avoid having an extra Thread, just actively waiting to shutdown ZK. Would this make sense? Or am I missing something maybe?

Reply:
Ah, I see, sorry I misunderstood. Yes, it appears that that thread is rather unnecessary, cannot remember why i put it... Thanks for the callout.